lib/internal/source_map/prepare_stack_trace.js

```diff
@@ -49,18 +59,50 @@ const prepareStackTrace = (globalThis, error, trace) => {
         } = sm.findEntry(t.getLineNumber() - 1, t.getColumnNumber() - 1);
         if (originalSource && originalLine !== undefined &&
             originalColumn !== undefined) {
-          str +=
-`\n        -> ${originalSource.replace('file://', '')}:${originalLine + 1}:${originalColumn + 1}`;
+          const originalSourceNoScheme = originalSource.replace('file://', '');
+          if (i === 0) {
+            firstLine = originalLine + 1;
+            firstColumn = originalColumn + 1;
+            firstSource = originalSourceNoScheme;
+          }
+          str += `\n        -> ${originalSourceNoScheme}:${originalLine + 1}:` +
+            `${originalColumn + 1}`;
         }
       }
     } catch (err) {
       debug(err.stack);
     }
     return str;
   });
-  return `${errorString}\n    at ${preparedTrace.join('')}`;
+  return `${getErrorSource(firstSource, firstLine, firstColumn)}` +
+          `${errorString}\n    at ${preparedTrace.join('')}`;
 };
 
+// Places a snippet of code from where the exception was originally thrown
+// above the stack trace. This logic is modeled after GetErrorSource in
+// node_errors.cc.
+function getErrorSource(firstSource, firstLine, firstColumn) {
+  let exceptionLine = '';
+  try {
```

Comment:
The try/catch block is way too big. It's only reasonable around the `fs.readFileSync()` call.

You'll want to `const { readFileSync } = require('fs');`' at the top to be immune to users monkey-patching fs.